ReplSeqMem Crashes on Register with Smaller Vec Init Trying to Pad Smaller Type

The use of `LegalizeConnects` inside `ReplSeqMem` can cause crashes if you have an aggregate register with an initializer that has a ground type smaller than the register.  `LegalizeConnects` then tries to pad the initializer, but it throws exceptions on any aggregate types.

The following is a minimal example:

```scala
circuit Foo:
  module Foo:
    input clock: Clock
    input reset: UInt<1>
    input a: UInt<1>[1]
    output b: UInt<2>[1]

    wire init: UInt<1>[1]
    init <= a

    reg r : UInt<2>[1], clock with :
      reset => (reset, init)

    b <= r
```

You _need to run `ReplSeqMem`s to get this to crash_:

```
# firrtl -i Foo.fir --repl-seq-mem -c:Foo:-o:Foo.mem.conf 
Exception in thread "main" firrtl.FirrtlInternalException: Internal Error! Please file an issue at https://github.com/ucb-bar/firrtl/issues
	at firrtl.Utils$.error(Utils.scala:503)
	at firrtl.Utils$.throwInternalError(Utils.scala:175)
	at firrtl.stage.phases.CatchExceptions.transform(CatchExceptions.scala:38)
	at firrtl.stage.phases.CatchExceptions.transform(CatchExceptions.scala:17)
	at firrtl.options.phases.DeletedWrapper.internalTransform(DeletedWrapper.scala:38)
	at firrtl.options.phases.DeletedWrapper.internalTransform(DeletedWrapper.scala:15)
	at firrtl.options.Translator.transform(Phase.scala:248)
	at firrtl.options.Translator.transform$(Phase.scala:248)
	at firrtl.options.phases.DeletedWrapper.transform(DeletedWrapper.scala:15)
	at firrtl.options.DependencyManager.$anonfun$transform$5(DependencyManager.scala:280)
	at firrtl.Utils$.time(Utils.scala:181)
	at firrtl.options.DependencyManager.$anonfun$transform$3(DependencyManager.scala:280)
	at scala.collection.LinearSeqOps.foldLeft(LinearSeq.scala:169)
	at scala.collection.LinearSeqOps.foldLeft$(LinearSeq.scala:165)
	at scala.collection.immutable.List.foldLeft(List.scala:79)
	at firrtl.options.DependencyManager.transform(DependencyManager.scala:269)
	at firrtl.options.DependencyManager.transform$(DependencyManager.scala:255)
	at firrtl.options.PhaseManager.transform(DependencyManager.scala:443)
	at firrtl.stage.FirrtlStage.run(FirrtlStage.scala:38)
	at firrtl.options.Stage$$anon$1.transform(Stage.scala:43)
	at firrtl.options.Stage$$anon$1.transform(Stage.scala:43)
	at firrtl.options.phases.DeletedWrapper.internalTransform(DeletedWrapper.scala:38)
	at firrtl.options.phases.DeletedWrapper.internalTransform(DeletedWrapper.scala:15)
	at firrtl.options.Translator.transform(Phase.scala:248)
	at firrtl.options.Translator.transform$(Phase.scala:248)
	at firrtl.options.phases.DeletedWrapper.transform(DeletedWrapper.scala:15)
	at firrtl.options.Stage.$anonfun$transform$5(Stage.scala:47)
	at scala.collection.LinearSeqOps.foldLeft(LinearSeq.scala:169)
	at scala.collection.LinearSeqOps.foldLeft$(LinearSeq.scala:165)
	at scala.collection.immutable.List.foldLeft(List.scala:79)
	at firrtl.options.Stage.$anonfun$transform$3(Stage.scala:47)
	at logger.Logger$.$anonfun$makeScope$2(Logger.scala:167)
	at scala.util.DynamicVariable.withValue(DynamicVariable.scala:59)
	at logger.Logger$.makeScope(Logger.scala:165)
	at firrtl.options.Stage.transform(Stage.scala:47)
	at firrtl.options.Stage.execute(Stage.scala:58)
	at firrtl.options.StageMain.main(Stage.scala:71)
	at firrtl.stage.FirrtlMain.main(FirrtlStage.scala)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at coursier.bootstrap.launcher.a.a(Unknown Source)
	at coursier.bootstrap.launcher.Launcher.main(Unknown Source)
Caused by: java.lang.RuntimeException: forceWidth does not support expressions of type VectorType(UIntType(IntWidth(1)),1)
	at firrtl.passes.PadWidths$.forceWidth(PadWidths.scala:36)
	at firrtl.passes.LegalizeConnects$.onStmt(LegalizeConnects.scala:24)
	at firrtl.passes.LegalizeConnects$.$anonfun$onStmt$1(LegalizeConnects.scala:25)
	at firrtl.ir.Block.mapStmt(IR.scala:653)
	at firrtl.passes.LegalizeConnects$.onStmt(LegalizeConnects.scala:25)
	at firrtl.passes.LegalizeConnects$.$anonfun$run$2(LegalizeConnects.scala:29)
	at firrtl.ir.Module.mapStmt(IR.scala:1198)
	at firrtl.passes.LegalizeConnects$.$anonfun$run$1(LegalizeConnects.scala:29)
	at scala.collection.immutable.List.map(List.scala:246)
	at scala.collection.immutable.List.map(List.scala:79)
	at firrtl.passes.LegalizeConnects$.run(LegalizeConnects.scala:29)
	at firrtl.passes.memlib.SimpleTransform.execute(ReplaceMemTransform.scala:112)
	at firrtl.Transform.runTransform(Compiler.scala:365)
	at firrtl.Transform.runTransform$(Compiler.scala:364)
	at firrtl.passes.memlib.SimpleTransform.runTransform(ReplaceMemTransform.scala:109)
	at firrtl.SeqTransformBased.$anonfun$runTransforms$1(Compiler.scala:374)
	at scala.collection.LinearSeqOps.foldLeft(LinearSeq.scala:169)
	at scala.collection.LinearSeqOps.foldLeft$(LinearSeq.scala:165)
	at scala.collection.immutable.List.foldLeft(List.scala:79)
	at firrtl.SeqTransformBased.runTransforms(Compiler.scala:374)
	at firrtl.SeqTransformBased.runTransforms$(Compiler.scala:373)
	at firrtl.SeqTransform.runTransforms(Compiler.scala:378)
	at firrtl.SeqTransform.execute(Compiler.scala:384)
	at firrtl.Transform.transform(Compiler.scala:296)
	at firrtl.Transform.transform$(Compiler.scala:296)
	at firrtl.SeqTransform.transform(Compiler.scala:378)
	at firrtl.stage.transforms.ExpandPrepares.execute(ExpandPrepares.scala:19)
	at firrtl.Transform.transform(Compiler.scala:296)
	at firrtl.Transform.transform$(Compiler.scala:296)
	at firrtl.stage.transforms.ExpandPrepares.transform(ExpandPrepares.scala:7)
	at firrtl.stage.transforms.CatchCustomTransformExceptions.execute(CatchCustomTransformExceptions.scala:10)
	at firrtl.Transform.transform(Compiler.scala:296)
	at firrtl.Transform.transform$(Compiler.scala:296)
	at firrtl.stage.transforms.CatchCustomTransformExceptions.transform(CatchCustomTransformExceptions.scala:7)
	at firrtl.stage.transforms.UpdateAnnotations.internalTransform(UpdateAnnotations.scala:22)
	at firrtl.stage.transforms.UpdateAnnotations.internalTransform(UpdateAnnotations.scala:8)
	at firrtl.options.Translator.transform(Phase.scala:248)
	at firrtl.options.Translator.transform$(Phase.scala:248)
	at firrtl.stage.transforms.UpdateAnnotations.transform(UpdateAnnotations.scala:8)
	at firrtl.options.DependencyManager.$anonfun$transform$5(DependencyManager.scala:280)
	at firrtl.Utils$.time(Utils.scala:181)
	at firrtl.options.DependencyManager.$anonfun$transform$3(DependencyManager.scala:280)
	at scala.collection.LinearSeqOps.foldLeft(LinearSeq.scala:169)
	at scala.collection.LinearSeqOps.foldLeft$(LinearSeq.scala:165)
	at scala.collection.immutable.List.foldLeft(List.scala:79)
	at firrtl.options.DependencyManager.transform(DependencyManager.scala:269)
	at firrtl.options.DependencyManager.transform$(DependencyManager.scala:255)
	at firrtl.stage.TransformManager.transform(TransformManager.scala:14)
	at firrtl.stage.phases.Compiler.$anonfun$internalTransform$6(Compiler.scala:138)
	at firrtl.Utils$.time(Utils.scala:181)
	at firrtl.stage.phases.Compiler.f$1(Compiler.scala:138)
	at firrtl.stage.phases.Compiler.$anonfun$internalTransform$8(Compiler.scala:143)
	at scala.collection.immutable.List.map(List.scala:246)
	at scala.collection.immutable.List.map(List.scala:79)
	at firrtl.stage.phases.Compiler.internalTransform(Compiler.scala:143)
	at firrtl.stage.phases.Compiler.internalTransform(Compiler.scala:53)
	at firrtl.options.Translator.transform(Phase.scala:248)
	at firrtl.options.Translator.transform$(Phase.scala:248)
	at firrtl.stage.phases.Compiler.transform(Compiler.scala:53)
	at firrtl.stage.phases.CatchExceptions.transform(CatchExceptions.scala:30)
	... 41 more
```

Everything is a-okay without `ReplSeqMem`s.

--- a/src/main/scala/firrtl/passes/LegalizeConnects.scala
+++ b/src/main/scala/firrtl/passes/LegalizeConnects.scala
@@ -29,3 +29,23 @@ object LegalizeConnects extends Pass {
     c.copy(modules = c.modules.map(_.mapStmt(onStmt)))
   }
 }
+
+/** Ensure that all connects have the same bit-width on the RHS and the LHS.
+  */
+private[firrtl] object LegalizeConnectsOnly extends Pass {
+
+  override def prerequisites = Seq(Dependency(ExpandConnects))
+  override def optionalPrerequisites = Seq.empty
+  override def optionalPrerequisiteOf = Seq.empty
+  override def invalidates(a: Transform) = false
+
+  def onStmt(s: Statement): Statement = s match {
+    case c: Connect =>
+      c.copy(expr = PadWidths.forceWidth(bitWidth(c.loc.tpe).toInt)(c.expr))
+    case other => other.mapStmt(onStmt)
+  }
+
+  def run(c: Circuit): Circuit = {
+    c.copy(modules = c.modules.map(_.mapStmt(onStmt)))
+  }
+}

--- a/src/main/scala/firrtl/passes/memlib/ReplaceMemTransform.scala
+++ b/src/main/scala/firrtl/passes/memlib/ReplaceMemTransform.scala
@@ -152,7 +152,7 @@ class ReplSeqMem extends SeqTransform with HasShellOptions with DependencyAPIMig
 
   val transforms: Seq[Transform] =
     Seq(
-      new SimpleMidTransform(LegalizeConnects),
+      new SimpleMidTransform(LegalizeConnectsOnly),
       new SimpleMidTransform(ToMemIR),
       new SimpleMidTransform(ResolveMaskGranularity),
       new SimpleMidTransform(RenameAnnotatedMemoryPorts),

--- a/src/test/scala/firrtlTests/ReplSeqMemTests.scala
+++ b/src/test/scala/firrtlTests/ReplSeqMemTests.scala
@@ -672,4 +672,25 @@ circuit Top :
     )
     resAnnos should be(expected)
   }
+
+  "ReplSeqMem" should "not crash if there are aggregate registers in the design that require padding (see #2379)" in {
+
+    val input =
+      """|circuit Foo:
+         |  module Foo:
+         |    input clock: Clock
+         |    input reset: UInt<1>
+         |    input a: UInt<1>[1]
+         |    output b: UInt<2>[1]
+         |
+         |    wire init: UInt<1>[1]
+         |    init <= a
+         |
+         |    reg r : UInt<2>[1], clock with :
+         |      reset => (reset, init)
+         |
+         |    b <= r
+         |""".stripMargin
+    compileAndEmit(CircuitState(parse(input), ChirrtlForm))
+  }
 }

